istresearch/dog-whistle
# Multi gauge support

Adds the ability to support multiple gauges to be configured for a single log statement. This allows a traditional dictionary of items to be parsed and sent to Datadog as separate gauges.

To Test:

1. Ensure DogWhistle `0.3.0` is installed in your environment
```
$ python setup.py install
$ pip freeze | grep dog
datadog==0.15.0
dog-whistle==0.3.0
```

2. Ensure you have the `scutils` library installed
```
$ pip install scutils -U --pre
```
The `--pre` flag allows you to use `1.3.0dev0`, which is required for python 2/3 compatibility

3. Run the statsd container as outlined in the docs
```
$ docker run -p 80:80 -p 2003-2004:2003-2004 -p 2023-2024:2023-2024 -p 8125:8125/udp -p 8126:8126 hopsoft/graphite-statsd
```

4. Run the example script located within the root of the project
```
$ python example.py
... # you should see something similar to the following
2017-05-17 15:20:10,182 [dog_whistle] DEBUG: executing gauge log to datadog
2017-05-17 15:20:10,183 [dog_whistle] INFO: metric guage scrapy_cluster.too_high
2017-05-17 15:20:10,183 [dog_whistle] INFO: metric guage scrapy_cluster.too_high_gauge2
```

5. Check your statsd host by navigating to [localhost](http://localhost), clicking `Metrics/stats/gauges/scrapy_cluster/`. You should see `too_high` and `too_high_gauge2` metrics.

6. Click both of the metrics to add them to your graph, and change the time range on the graph to be 5 mins. You should see two lines on your graph like below

<img width="492" alt="screen shot 2017-05-17 at 3 26 41 pm" src="https://cloud.githubusercontent.com/assets/3504194/26172118/41606b96-3b15-11e7-92ac-1b52115a0bdc.png">

The important part here is that both metrics are displayed, even though in the code we only called the log statement one time like so:

```
logger.error("Too high!", {'wrapper': {'value': val}, 'key2': 11})
```

You can rerun step 4 to generate slight variations in the `too_high` graph to get an output like above.

If you made it this far, this completes the new capability to add multiple gauges to a single log statement.
